dhash/module.py

```diff
@@ -0,0 +1,508 @@
+                        )
+                        + "\n"
+                        + _(ctx, "Calculated **{hashes}** hashes.")
+                    ).format(
+                        count=i,
+                        total=len(messages),
+                        percent="{:.1f}".format(i / len(messages) * 100),
+                        hashes=ctr_hashes,
+                    )
+                )
+
+            if not len(message.attachments):
+                ctr_nofile += 1
+                continue
+
+            hashes = [x async for x in self._save_hashes(message)]
+            ctr_hashes += len(hashes)
+
+        await status.edit(
+            content=(
+                _(ctx, "**COMPLETED**")
+                + "\n"
+                + _(ctx, "Processed **{messages}** messages.")
+                + "\n"
+                + _(
+                    ctx,
+                    "Calculated **{hashes}** image hashes in **{seconds}** seconds.",
+                )
+            ).format(
+                messages=len(messages),
+                hashes=ctr_hashes,
+                seconds="{:.1f}".format(time.time() - now),
+            )
+        )
+
+    @commands.check(check.acl)
+    @dhash.command(name="compare", aliases=["messages"])
+    async def dhash_compare(self, ctx, messages: commands.Greedy[discord.Message]):
+        """Display hashes of given messages.
+        messages: Space separated list of messages.
+        """
+        text = []
+
+        for message in messages:
+            db_images = ImageHash.get_by_message(message.guild.id, message.id)
+            if not len(db_images):
+                continue
+
+            text.append(
+                _(ctx, "Message **`{message_id}`**").format(message_id=message.id)
+            )
+            for db_image in db_images:
+                text.append("   > `{hash}`".format(hash=db_image.dhash[2:]))
+            text.append("")
+
+        if not len(text):
+            return await ctx.send(_(ctx, "Messages has no associtated hashes."))
+
+        await ctx.send("\n".join(text))
+
+    @commands.Cog.listener()
+    async def on_message(self, message: discord.Message):
+        if self._in_repost_channel(message):
+            await self._check_message(message)
+
+    @commands.Cog.listener()
+    async def on_raw_message_delete(self, payload: discord.RawMessageDeleteEvent):
+        channel = HashChannel.get(payload.guild_id, payload.channel_id)
+        if channel:
+            ImageHash.delete_by_message(payload.guild_id, payload.message_id)
+
+    @commands.Cog.listener()
+    async def on_message_delete(self, message: discord.Message):
+        if not self._in_repost_channel(message):
+            return
+
+        if message.author == self.bot.user:
+            return
+
+        if message.id in self.embed_cache:
+            try:
+                report = await message.channel.fetch_message(
```

Comment:
We have a function to fetch messages only when necessary:
```py
report = await utils.Discord.get_message(...)
```
It will check the bot cache first before requesting the message via API.

Reply:
Actually we don't need that since we use cache for this now.